torch/testing/_internal/common_distributed.py

```diff
@@ -247,8 +247,33 @@ def _run(cls, rank, test_name, file_name):
     def _join_processes(self, fn):
         timeout = get_timeout(self.id())
         start_time = time.time()
-        for p in self.processes:
-            p.join(timeout)
+        subprocess_error = False
+        while True:
+            # check to see if any subprocess exited with an error early.
+            for p in self.processes:
+                # This is the exited code processes exit with if they encountered an exception.
+                if p.exitcode == MultiProcessTestCase.TEST_ERROR_EXIT_CODE:
+                    print("Some process exited badly, terminating rest.")
+                    active_children = torch.multiprocessing.active_children()
+                    for ac in active_children:
+                        ac.terminate()
+                    subprocess_error = True
+            if subprocess_error:
+                break
+            # All processes have joined cleanly if they all a valid exitcode
+            if all([p.exitcode is not None for p in self.processes]):
+                break
+            # Check if we should time out the test. If so, we terminate each process.
+            elapsed = time.time() - start_time
+            if elapsed > timeout:
+                print(
+                    "Timing out after {} seconds and killing subprocesses.".format(
+                        timeout
+                    )
+                )
+                for p in self.processes:
+                    p.terminate()
```

Comment:
will this be a no-op for errored processes? Because there could be race that some process crashed after we check their error code above.

Reply:
Yes, it looks like it is safe to call `terminate()` multiple times/on an already-exited process. Although, maybe we could use `multiprocessing.active_children()`.